---
title: "SingleCell_Kurten"
output: html_document
---

```{r}
library(Seurat)
library(SeuratDisk)
library(tidyverse)
library(ggplot2)
library(gridExtra)
library(cowplot)
```

# Intregate single-cell RNA-Seq dataset in R using Seurat for Kürten Data

In this workflow we're going to use the scRNA data in Kürten publication (DOI: https://doi.org/10.1038/s41467-021-27619-4). The data was downloaded from the GEO platform on the 16th of May of 2022 (accession number: GSE164690).

We can first get the names of the directories that contain our files:

```{r}
dirs <- list.dirs(path = '~/Single_Cell_Analysis/GSE164690_RAW', recursive = F, full.names = F)
```

```{r}
dirs
```

```{r}
gsub("^[^_]*_","", "GSM5017021_HN01_PBL")
```

To upload the data for each patient and type of sample as a Seurat object we create a loop in which we first create a Mtx object with the matrix, feature and cell files and from this we create a Seurat object for each.

```{r}
for (dir in dirs[26:51]){
  name <- gsub("^[^_]*_","", dir)
  cts <- ReadMtx( mtx =
                paste0('~/Single_Cell_Analysis/GSE164690_RAW/', dir, 
                       '/matrix.mtx.gz'), features = 
                paste0('~/Single_Cell_Analysis/GSE164690_RAW/', dir, 
                       '/features.tsv.gz'), cells = 
                paste0('~/Single_Cell_Analysis/GSE164690_RAW/', dir, 
                       '/barcodes.tsv.gz'))
  assign(name, CreateSeuratObject(counts = cts))
}
```

```{r}
names <- c()
for (dir in dirs){
names <- c(names, gsub("^[^_]*_","", dir))
}
```

```{r}
names <- names[-25]
```

```{r}
cts <- ReadMtx( mtx =
                paste0('~/Single_Cell_Analysis/GSE164690_RAW/GSM5017045_HN10_PBL/matrix.mtx.gz'), features = 
                paste0('~/Single_Cell_Analysis/GSE164690_RAW/GSM5017045_HN10_PBL/features.tsv.gz'), cells = 
                paste0('~/Single_Cell_Analysis/GSE164690_RAW/GSM5017045_HN10_PBL/barcodes.tsv.gz'))
  assign("HN10_PBL", CreateSeuratObject(counts = cts))
```

## Quality Control of the Data

This could be done for each of the 51 objects individually, but this process will be more time consuming and error prone since we will have to do it manually. Then we can merge all the datasets into one, this way all will be processed at the same time in the same way.

```{r}
merged_seurat <- merge(HN01_PBL, y = c(HN01_CD45p, HN01_CD45n, HN02_PBL, HN02_CD45p, HN03_PBL,   HN03_CD45p, HN04_PBL, HN04_CD45p, HN05_PBL, HN05_CD45p, HN05_CD45n, HN06_PBL, HN06_CD45p, HN06_CD45n,HN07_PBL, HN07_CD45p, HN07_CD45n, HN08_PBL, HN08_CD45p, HN08_CD45n, HN09_PBL,HN09_CD45p, HN09_CD45n, HN10_CD45p, HN10_CD45n, HN11_PBL, HN11_CD45p, HN11_CD45n,HN12_PBL, HN12_CD45p, HN12_CD45n, HN13_PBL,HN13_CD45p, HN13_CD45n, HN14_PBL,HN14_CD45p, HN14_CD45n, HN15_PBL, HN15_CD45p, HN15_CD45n, HN16_PBL, HN16_CD45p,HN16_CD45n, HN17_PBL,  HN17_CD45p, HN17_CD45n, HN18_PBL, HN18_CD45p, HN18_CD45n), add.cell.ids = names, project = "HNSC")
```

```{r}
merged_seurat
```

### QC and filtering

```{r}
rownames(merged_seurat@meta.data)[1:10]
```

We see that each sample has a prefix which indicates the patient and origin of the sample, for each patient we have:
- CD45p: immune cells from and  freshly resected HNSCC tumors
- CD45n: epithelial and stromal cells from freshly resected HNSCC tumors
- PBL:  matched peripheral blood leukocytes

So we can start by creating a column in the metadata dataset that has this information:

```{r}
merged_seurat$sample <- rownames(merged_seurat@meta.data)
merged_seurat$sample_code <- sub("^(([^_]*_){1}[^_]*).*", "\\1", rownames(merged_seurat@meta.data))
```

```{r}
merged_seurat@meta.data <- separate(merged_seurat@meta.data, col = "sample", into = c("Patient", "Type", "Barcode"), sep = '_')
```

```{r}
clinical_data <- data.frame(Sex = c("male","female", "male", "male", "female", "male", "male", "female", "female", "male", "male", "male", "male","male", "female", "male", "male", "male"), Age = as.factor(c("70-79","60-69","80-89", "50-59", "50-59", "30-39", "60-69", "70-79","70-79","50-59","80-89", "50-59","70-79","50-59", "60-69", "40-49", "50-59", "50-59")), Smoking = c("Yes", "No", "No", "Yes","Yes","Yes","Yes","Yes","Yes","No", "No", "Yes", "No", "Yes","Yes","Yes","Yes","Yes"), Alcohol = c("No", "No", "No", "Yes","Yes","Yes","Yes","Yes","Yes","Yes","No", "Yes","No", "Yes", NA, "Yes","Yes","Yes"), Disease_site = c("Oral cavity","Oral cavity","Oral cavity","Oral cavity","Oral cavity","Oral cavity","Larynx","Oral cavity","Oral cavity","Oral cavity","Oral cavity","Oropharynx","Oropharynx","Oropharynx","Oral cavity","Oropharynx","Oropharynx","Oropharynx"), HPV16 = c("Neg","Neg","Neg","Neg","Neg","Neg","Neg","Neg","Neg","Neg","Neg","Pos","Pos","Pos","Neg","Pos","Pos","Pos"), Inflamation_status = c("High", "Low", NA, "Low", "Med", "High", "Low", "Med", "Med", "High", "Low", "Med", "High", "High", "Low", "Med", "High", "Low"))

```

```{r}
patient <- unique(merged_seurat@meta.data$Patient)
rownames(clinical_data) <- patient
```

We write a table with the clinical data:

```{r}
write_csv(clinical_data, "/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/SC_HNSCC_clinical_data.csv")
```

#### Percentage of mitochondrial genes:

```{r}
merged_seurat$mitoPercent <- PercentageFeatureSet(merged_seurat, pattern = '^MT-')
```


#### Visualizing the number of cells detected

Aimed to recover 2000 cells each of tumor-infiltrating lymphocytes (TIL, CD45+), TME non-immune cells (i.e., epithelial and stromal cells, CD45-) and PBL from each patient.

```{r}
merged_seurat@meta.data %>% 
  	ggplot(aes(x=sample_code, fill=sample_code)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")
```

We can see that some of the samples have a lot of "bad" cells, reads that come from droplets which only had the barcode present or similar.

#### Complexity of the data, number of genes detected per UMI

- Reads with different Unique Molecular Identifiers (UMIs) mapping to the same transcript were derived from different molecules and are biological duplicates each read should be counted.
- Reads with the same UMI originated from the same molecule and are technical duplicates the UMIs should be collapsed to be counted as a single read.

We take the log10 of the number of genes detected per cell and the log10 of the number of UMIs per cell, then divide the log10 number of genes by the log10 number of UMIs.

```{r}
merged_seurat$log10GenesPerUMI <- log10(merged_seurat$nFeature_RNA) / log10(merged_seurat$nCount_RNA)
```

The UMI counts per cell should generally be above 500, that is the low end of what we expect. If UMI counts are between 500-1000 counts, it is usable but the cells probably should have been sequenced more deeply.

```{r}
# Visualize the number UMIs/transcripts per cell
merged_seurat@meta.data %>% 
  	ggplot(aes(color=sample_code, x=nCount_RNA, fill= sample_code)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)
```

In this case the samples are all over 1000 so the quality of sequencing depth is pretty good.

#### Genes detected per cell

We have similar expectations for gene detection as for UMI detection, although it may be a bit lower than UMIs. For high quality data, the proportional histogram should contain a single large peak that represents cells that were encapsulated. If we see a small shoulder to the left of the major peak (not present in our data), or a bimodal distribution of the cells, that can indicate a couple of things. It might be that there are a set of cells that failed for some reason. It could also be that there are biologically different types of cells (i.e. quiescent cell populations, less complex cells of interest), and/or one type is much smaller than the other (i.e. cells with high counts may be cells that are larger in size). Therefore, this threshold should be assessed with other metrics that we describe in this lesson.


```{r}
# Visualize the distribution of genes detected per cell via histogram
merged_seurat@meta.data %>% 
  	ggplot(aes(color=sample_code, x=nFeature_RNA, fill= sample_code)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)

# Visualize the distribution of genes detected per cell via boxplot
merged_seurat@meta.data %>% 
  	ggplot(aes(x=sample_code, y=log10(nFeature_RNA), fill=sample_code)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells vs NGenes")
```

All samples seem to be okay except for sample HN15-CD45-, should we remove?

#### UMIs vs. genes detected

Two metrics that are often evaluated together are the number of UMIs and the number of genes detected per cell. Here, we have plotted the number of genes versus the number of UMIs coloured by the fraction of mitochondrial reads. Mitochondrial read fractions are only high in particularly low count cells with few detected genes (darker colored data points). This could be indicative of damaged/dying cells whose cytoplasmic mRNA has leaked out through a broken membrane, and thus, only mRNA located in the mitochondria is still conserved. These cells are filtered out by our count and gene number thresholds. Jointly visualizing the count and gene thresholds shows the joint filtering effect.

Cells that are poor quality are likely to have low genes and UMIs per cell, and correspond to the data points in the bottom left quadrant of the plot. Good cells will generally exhibit both higher number of genes per cell and higher numbers of UMIs.

With this plot we also evaluate the slope of the line, and any scatter of data points in the bottom right hand quadrant of the plot. These cells have a high number of UMIs but only a few number of genes. These could be dying cells, but also could represent a population of a low complexity celltype (i.e red blood cells).

```{r}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
merged_seurat@meta.data %>% 
  	ggplot(aes(x=nFeature_RNA, y=nCount_RNA, color=mitoPercent)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 200) +
  	facet_wrap(~sample_code)
```

```{r}
VlnPlot(merged_seurat, features = c("nFeature_RNA","nCount_RNA", "mitoPercent"))
```

We can look at this features together in the different axis:

```{r}
FeatureScatter(merged_seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") +
  geom_smooth(method = "lm")
```
#### Mitochondrial counts ratio

This metric can identify whether there is a large amount of mitochondrial contamination from dead or dying cells. We define poor quality samples for mitochondrial counts as cells which surpass the 0.2 mitochondrial ratio mark, unless of course you are expecting this in your sample.

```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
merged_seurat@meta.data %>% 
  	ggplot(aes(color=sample_code, x=mitoPercent, fill=sample_code)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)
```

#### Complexity

We can evaluate each cell in terms of how complex the RNA species are by using a measure called the novelty score. The novelty score is computed by taking the ratio of nGenes (nFeature_RNA) over nUMI (nCounts_RNA). If there are many captured transcripts (high nUMI) and a low number of genes detected in a cell, this likely means that you only captured a low number of genes and simply sequenced transcripts from those lower number of genes over and over again. These low complexity (low novelty) cells could represent a specific cell type (i.e. red blood cells which lack a typical transcriptome), or could be due to some other strange artifact or contamination. Generally, we expect the novelty score to be above 0.80 for good quality cells.

```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
merged_seurat@meta.data %>%
  	ggplot(aes(x=log10GenesPerUMI, color = sample_code, fill=sample_code)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)
```


Based on the QC metrics suggested in the Scanpy tutorial(doi: 10.1186/s13059-017-1382-0), cells with less than 200 genes expressed were filtered out. Cells expressing more than 5000 genes, and more than ten percent mitochondrial genes were also removed. In this case the sequencing depth, number of transcripts per gene seems to be okay since there're no dots in the top left corner of the graph so this parameter is not going to be touched.

```{r}
merged_seurat_filtered <- subset(merged_seurat, subset = nCount_RNA > 200 & nCount_RNA < 5000 & mitoPercent < 10)
```

Genes expressed in less than 3 cells were also filtered out of the analysis:

```{r}
selected_f <- rownames(merged_seurat_filtered)[Matrix::rowSums(merged_seurat_filtered) > 3]
merged_seurat_filtered <- subset(merged_seurat_filtered, features = selected_f)
```

```{r}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
merged_seurat_filtered@meta.data %>% 
  	ggplot(aes(x=nFeature_RNA, y=nCount_RNA, color=mitoPercent)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black", limits=c(0,100)) +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 200) +
  	facet_wrap(~sample_code)
```

# Normalization and regressing out unwanted variation

#### Batch effect

Since we have samples from different sources is important to check if we see any batch effects:

1. Normalize data:

```{r}
merged_seurat_filtered <- NormalizeData(object = merged_seurat_filtered)
```

#### Evaluating effects of cell cycle

To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seurat function CellCycleScoring(). This function calculates cell cycle phase scores based on canonical markers that required as input. We retrieve S and G2M phase from that are uploaded with Seurat.

```{r}
s_genes <- cc.genes$s.genes
g2m_genes <- cc.genes$g2m.genes
```

```{r}
# Score cells for cell cycle
merged_seurat_filtered <- CellCycleScoring(merged_seurat_filtered, 
                                 g2m.features = g2m_genes, 
                                 s.features = s_genes)

# View cell cycle scores and phases assigned to cells                                 
View(merged_seurat_filtered@meta.data) 
```



2. Find variable features:

```{r}
merged_seurat_filtered <-  FindVariableFeatures(object = merged_seurat_filtered)
```

3. Scale Data:

```{r}
merged_seurat_filtered <-  ScaleData(object = merged_seurat_filtered)
```

4. Run PCA Analysis

```{r}
merged_seurat_filtered <- RunPCA(object = merged_seurat_filtered)
```

```{r}
# Plot the PCA colored by cell cycle phase
DimPlot(merged_seurat_filtered,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase")
```

It doesn't seem that there's a clusterization of cells based on it's cell cycle, so we're not going to account for that.

# Evaluating effects of mitochodrial expression

Mitochondrial expression is another factor which can greatly influence clustering. Oftentimes, it is useful to regress out variation due to mitochondrial expression. However, if the differences in mitochondrial gene expression represent a biological phenomenon that may help to distinguish cell clusters, then we advise not regressing this out. 

```{r}
# Check quartile values
summary(merged_seurat_filtered@meta.data$mitoPercent)
```

```{r}
# Turn mitoRatio into categorical factor vector based on quartile values
merged_seurat_filtered@meta.data$mitoFr <- cut(merged_seurat_filtered@meta.data$mitoPercent, 
                   breaks=c(-Inf, 1.924, 3.452, 4.655, Inf), 
                   labels=c("Low","Medium","Medium high", "High"))
```

```{r}
# Plot the PCA colored by mitochondrial fraction
DimPlot(merged_seurat_filtered,
        reduction = "pca",
        group.by= "mitoFr",
        split.by = "mitoFr")
```
The data doesn't seem to be grouping by the mitochondrial fraction.

In the case a variable was interceding we can account for it in the following step.

Now we can use the sctransform method as a more accurate method of normalizing, estimating the variance of the raw filtered data, and identifying the most variable genes. The sctransform method models the UMI counts using a regularized negative binomial model to remove the variation due to sequencing depth (total nUMIs per cell), while adjusting the variance based on pooling information across genes with similar abundances (similar to some bulk RNA-seq methods). Sctransform automatically accounts for cellular sequencing depth by regressing out sequencing depth (nUMIs). However, if there are other sources of uninteresting variation identified in the data during the exploration steps we can also include these. 

```{r}
saveRDS(merged_seurat_filtered, "merged_seurat_filtered.rds")
```

```{r, eval = F}
# this step we perform in the cluster because is computationally too heavy so we submit as an R file
#!/software/r/4.0.0/bin/Rscript
#BSUB -q medium
#BSUB -W 1:00 
#BSUB -n 2 
#BSUB -R span[hosts=1]
#BSUB -R "rusage[mem=100G]" 
#BSUB -J SCTransform
#BSUB -u cristina.condelopez@dkfz-heidelberg.de

library(Seurat)

merged_seurat_filtered <-readRDS("/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/merged_seurat_filtered.rds")

merged_seurat_filtered <- SCTransform(merged_seurat_filtered)

saveRDS(merged_seurat_filtered, "/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/merged_seurat_filtered_transform.rds")
```

```{r}
merged_seurat_filtered <-readRDS("/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/merged_seurat_filtered.rds")
```

```{r}
merged_seurat_filtered_transform <-readRDS("/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/merged_seurat_filtered_transform.rds")
```


```{r}
# Now, we run the following loop to perform the sctransform on all samples. This may take some time (~10 minutes):

for (i in 1:length(split_seurat)) {
    split_seurat[[i]] <- SCTransform(split_seurat[[i]], vars.to.regress = c("mitoRatio"))
    }
```


5. Find the Dimensions of the Dataset:

```{r}
ElbowPlot(merged_seurat_filtered_transform)
```

We can see that the variation seems to be captured in the first 15 components. We are going to use all the 20 components for the next steps:

```{r}
merged_seurat_filtered_transform <- FindNeighbors(object = merged_seurat_filtered_transform, dims = 1:20)
```

```{r}
DefaultAssay(merged_seurat_filtered_transform)
```


We clusterize the data:

```{r}
merged_seurat_filtered_transform <- FindClusters(object = merged_seurat_filtered_transform, graph.name = "RNA_snn")
```

And now non-linear dimensionality reduction:

```{r}
merged_seurat_filtered_transform <- RunUMAP(object = merged_seurat_filtered_transform, dims = 1:20)
```

Now we're going to plot the data. We want to know which cells came from each patient so we color them together, also by sample type.

```{r}
p1 <- DimPlot(merged_seurat_filtered_transform, reduction = "umap", group.by = "Patient")
```

```{r}
p2 <- DimPlot(merged_seurat_filtered_transform, reduction = "umap", group.by = "Type", cols = c("red", "green", "blue"))
```

```{r}
g <- grid.arrange(p1,p2, ncol =2, nrow = 2)
ggsave(file="non_integrated.pdf", g, width = 40,
  height = 30, units = c("cm"))
```

##### Perform integration to correct for batch effects


```{r}
obj.list <- SplitObject(merged_seurat_filtered, split.by = "Patient")
```

```{r}
for (i in 1:length(obj.list)){
  obj.list[[i]] <- SCTransform(obj.list[[i]])
  obj.list[[i]] <- FindVariableFeatures(object = obj.list[[i]])
}
```

We now select integration features:

```{r}
features <- SelectIntegrationFeatures(object.list = obj.list )
```


```{r}
# Prepare the SCT list object for integration
obj.list <- PrepSCTIntegration(object.list =  obj.list, 
                                   anchor.features = features)
```

```{r}
saveRDS(obj.list, "/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/obj_list.rds")
```

When we have selected the integration features we have to find the integration anchors (CCA), to integrate the data across different patients. In order to do that we are going to utilize a method called Canonical Correlation Analysis. Since this step is computationally heavy we're going to perform it on the cluster:

```{r, eval=FALSE}
#!/software/r/4.0.0/bin/Rscript
#BSUB -q long
#BSUB -W 5:00 
#BSUB -n 2 
#BSUB -R span[hosts=1]
#BSUB -R "rusage[mem=100G]" 
#BSUB -J AnchorsIntegration
#BSUB -u cristina.condelopez@dkfz-heidelberg.de

library(Seurat)

obj.list <-readRDS("/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/obj_list.rds")

anchors <- FindIntegrationAnchors(object.list = obj.list, normalization.method = "SCT", anchor.features = features)

seurat.integrated <- IntegrateData(anchorset = anchors,
                                   normalization.method = "SCT")

saveRDS(seurat.integrated, "/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/seurat_integrated.rds")
```

```{r}
anchors <- FindIntegrationAnchors(object.list = obj.list, normalization.method = "SCT", anchor.features = features)
```

And with this we can now integrate the data:

```{r}
seurat.integrated <- IntegrateData(anchorset = anchors,
                                   normalization.method = "SCT")
```



```{r}
seurat_integrated <-readRDS("/omics/odcf/analysis/OE0509_projects/hnscc/single_cell/Single_Cell_Analysis/seurat_integrated.rds")
```

Scale data
```{r}
seurat_integrated <- ScaleData(seurat_integrated)
```

```{r}
seurat_integrated <- RunPCA(seurat_integrated)
```

```{r}
seurat_integrated <- RunUMAP(seurat_integrated, dims = 1:50)
```

```{r}
p3 <- DimPlot(seurat_integrated, reduction = "umap", group.by = "Patient")
p4 <- DimPlot(seurat_integrated, reduction = "umap", group.by = "Type")
```

```{r}
g <- grid.arrange(p3, p4, ncol = 2)
```

```{r}
g <- grid.arrange(p3, p4, ncol = 2)
ggsave(file="integrated.pdf", g, width = 40,
  height = 20, units = c("cm"))
```

#### Identify significant PCs

```{r}
DimHeatmap(seurat_integrated, 
           dims = 1:15, 
           cells = 500, 
           balanced = TRUE)
```

```{r}
ElbowPlot(object = seurat_integrated, 
          ndims = 50)
```

```{r}
# Determine the K-nearest neighbor graph
seurat_integrated <- FindNeighbors(object = seurat_integrated, 
                                dims = 1:50)
```

```{r}
# Determine the clusters for various resolutions                                
seurat_integrated <- FindClusters(object = seurat_integrated,
                               resolution = c(0.4, 0.6, 0.8, 1.0, 1.4))
```

```{r}
# Explore resolutions
seurat_integrated@meta.data %>% 
        View()
```

```{r}
DimPlot(seurat_integrated, group.by = "integrated_snn_res.0.4", label = T)
DimPlot(seurat_integrated, group.by = "integrated_snn_res.0.6", label = T)
DimPlot(seurat_integrated, group.by = "integrated_snn_res.0.8", label = T)
DimPlot(seurat_integrated, group.by = "integrated_snn_res.1", label = T)
DimPlot(seurat_integrated, group.by = "integrated_snn_res.1.4", label = T)
```

```{r}
# Assign identity of clusters
Idents(object = seurat_integrated) <- "integrated_snn_res.1"
```

```{r}
metrics <-  c("nCount_RNA", "nFeature_RNA", "S.Score", "G2M.Score", "mitoPercent")

FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
```

```{r}
# Defining the information in the seurat object of interest
columns <- c(paste0("PC_", 1:16),
            "ident",
            "UMAP_1", "UMAP_2")

# Extracting this data from the seurat object
pc_data <- FetchData(seurat_integrated,  vars = columns) 
      
```

```{r}
# Adding cluster label to center of cluster on UMAP
umap_label <- FetchData(seurat_integrated, 
                        vars = c("ident", "UMAP_1", "UMAP_2"))  %>%
  group_by(ident) %>%
  summarise(x=mean(UMAP_1), y=mean(UMAP_2))
  
# Plotting a UMAP plot for each of the PCs
map(paste0("PC_", 1:16), function(pc){
        ggplot(pc_data, 
               aes(UMAP_1, UMAP_2)) +
                geom_point(aes_string(color=pc), 
                           alpha = 0.7) +
                scale_color_gradient(guide = FALSE, 
                                     low = "grey90", 
                                     high = "blue")  +
                geom_text(data=umap_label, 
                          aes(label=ident, x, y)) +
                ggtitle(pc)
}) %>% 
        plot_grid(plotlist = .)
```

```{r}
DimPlot(seurat_integrated, reduction = "umap", group.by = "ident", split.by = "Type", label = T)
```


### Exploring known cell type markers

For this we use the one described in the paper.

![](markers.png) 

**CD4+ T-cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD3D", "IL7R"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

Overlapping of the two 16/23?

**Monocytes**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD14", "LYZ"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

Overlapping of the two 9/0?

**CD8+ T cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD3D", "CD8A"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

Cluster 4/5/12?

**Reg T-cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD3D", "FOXP3"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```
Cluster 16/8?

**B cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD79A", "CD79B"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

Cluster 18/13

**Macrophagues**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD14", "CD68", "LYZ"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

Cluster 24/10?


**NK Cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("KLRD1"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

Cluster 1/20/5?

**Endothelial Cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CLDN5", "RAMP2"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

**Fibroblasts**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("COL1A1", "COL1A2", "DCN"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```
22?

**Ephitelial cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("KRT14", "KRT17"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

**Monocyte**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("MS4A7", "CD68"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```
28/10?

**Dendritic Cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("CD83", "CD86", "LYZ"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

28/9/0/10/24?

**pDendritic Cells**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("IL3RA", "CD68"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

**pericytes**

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c("COL1A1", "COL1A2", "RGS5"), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

```{r}
FeaturePlot(seurat_integrated, 
            reduction = "umap", 
            features = c(""), 
            order = TRUE,
            min.cutoff = 'q10', 
            label = TRUE)
```

```{r}
Markers <- FindAllMarkers(seurat_integrated,
               # Minimal logfc of the gene in the cluster we want to find vs. in others.
               logfc.threshold = 0.25,
               # minumun frequency in which genes are detected in the populations
               min.pct = 0.1,
               # Only markers that are upregulated
               only.pos = TRUE,
               # type of DEanalysis we're using
               test.use = 'wilcox')
# Not necessary to select a cluster because it iteratively compares one cluster to all the other clusters.
```

```{r}
write_csv(Markers, "cluster_markers.csv")
```

```{r}
markers_list <- split.data.frame(Markers, Markers$cluster)
```


```{r}
features <- c("CD3D", "CD8A", "CD4", "IL7R", "FOXP3", "CD19", "CD79B", "IL3RA", "NCAM1", "NCR1", "KLRD1", "IL3RA", "CSF1R", "MSR1R", "CD14", "MS4A7", "CD68", "CD83", "FCGR1A", "LAMP3", "FPR1", "THBD", "CD80", "CD86", "CD1C", "CD209", "LYZ", "CLDN5", "FLT1", "CDH5", "RAMP2", "EPCAM", "KRT14", "KRT17", "COL1A1", "DCN", "COL1A2", "RGS5")
```

```{r}
data <- as.matrix(seurat_integrated@assays$integrated@scale.data)
```

```{r}
data2 <- as.matrix(seurat_integrated@assays$RNA@counts@i)
```

```{r}
which(features %in% rownames(data2))
```

```{r}
i <- which(rownames(Markers) %in% features)
```

```{r}
DoHeatmap(subset(seurat_integrated, downsample = 100),
  features = rownames(Markers)[i],
  cells = NULL,
  group.by = "ident",
  group.bar = TRUE,
  group.colors = NULL,
  disp.min = -2.5,
  disp.max = NULL,
  slot = "scale.data",
  assay = NULL,
  label = TRUE,
  size = 5.5,
  hjust = 0,
  angle = 45,
  raster = TRUE,
  draw.lines = TRUE,
  lines.width = NULL,
  group.bar.height = 0.02,
  combine = TRUE
)

```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `0` = 'Monocytes')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `1` = 'NK Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `10` = 'Macrophagues')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `11` = 'CD8+ T Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `12` = 'CD8+ T Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `13` = 'B-Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `14` = 'Epithelial Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `15` = 'Endothelial Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `16` = 'T Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `17` = 'MAIT Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `18` = 'B-Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `19` = 'B-Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `2` = 'T Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `20` = 'NK Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `21` = 'Erythroid Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `22` = 'Fibroblast')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `23` = 'Epithelial Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `24` = 'Macrophagues')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `25` = 'Epithelial Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `26` = 'Epithelial Cells')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `28` = 'Macrophagues')
```

```{r}
seurat_integrated <- RenameIdents(seurat_integrated, `29` = 'Macrophagues')
```

```{r}
chrY.gene <- c("DDX3Y",	"EIF1AY",	"KDM5D",	"NLGN4Y",	"RPS4Y1",	"SRY",	"TBL1Y",	"TMSB4Y",	"USP9Y",	"UTY",	"ZFY")
```

```{r}
i <- which(rownames(merged_seurat_filtered_transform@assays$RNA@counts) %in% chrY.gene)
```

```{r}
Y_counts <- merged_seurat_filtered_transform@assays$RNA@counts[i,]
```

```{r}
as_matrix <- function(mat){

  tmp <- matrix(data=0L, nrow = mat@Dim[1], ncol = mat@Dim[2])
  
  row_pos <- mat@i+1
  col_pos <- findInterval(seq(mat@x)-1,mat@p[-1])+1
  val <- mat@x
    
  for (i in seq_along(val)){
      tmp[row_pos[i],col_pos[i]] <- val[i]
  }
    
  row.names(tmp) <- mat@Dimnames[[1]]
  colnames(tmp) <- mat@Dimnames[[2]]
  return(tmp)
}
```

```{r}
n <- colSums(as_matrix(merged_seurat_filtered_transform@assays$RNA@counts[i,])) 
```

```{r}
m <- colSums(as_matrix(merged_seurat_filtered_transform@assays$RNA@counts))
```

```{r}
merged_seurat_filtered_transform$pct_chrY <- n/m
```

```{r}
VlnPlot(merged_seurat_filtered_transform, features =  "pct_chrY", group.by = "Patient")
```

```{r}
patient <- unique(merged_seurat_filtered_transform@meta.data$Patient)
rownames(clinical_data) <- patient
```

```{r}
clinical_data_male <- clinical_data[clinical_data$Sex == "male",]
```

```{r}
rownames(clinical_data_male)
```

```{r}
merged_seurat_filtered_transform_males <- subset(merged_seurat_filtered_transform, subset = Patient == "HN01" | Patient == "HN03" | Patient == "HN04" | Patient == "HN06" | Patient == "HN07" | Patient == "HN10" | Patient == "HN11" | Patient == "HN12" | Patient == "HN13" | Patient == "HN14" | Patient == "HN16" | Patient == "HN17" | Patient == "HN18")
```

```{r}
VlnPlot(merged_seurat_filtered_transform_males, features =  "pct_chrY", group.by = "Patient")
```
