---
title: "Getting TPM counts from TCGA"
author: "Cristina Conde Lopez"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
  prettydoc::html_pretty:
    toc: yes
    theme: hpstr
    highlight: github
    number_sections: yes
editor_options:
  chunk_output_type: inline
---

For this kind of job you need:

- Table of counts from RNA seq analysis from the samples of interest.
- GTF file to where the samples were mapped during the counting process.

## Get count data

```{r}
TCGA_counts <- read.csv("TCGA_counts.csv", row.names = 1)
```

## Get TPM data

First we need the GTF file, for this we use the one for GRCh38.p0 since is the one the TCGA counts are mapped to (ENSG). We download this GTF from the following pathway on the 13th of January 2022:

https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.annotation.gtf.gz

```{r, message}
library(GenomicFeatures)
```

We make a TxDb from GTF file:

```{r}
txdb <- makeTxDbFromGFF("gencode.v38.annotation.gtf")G
```

We get the information for the genes in this file:

```{r}
all.genes <- genes(txdb)
```

We import the list of ENSEMBL ID of interest of our data, which corresponds to the rownames of the TCGA_counts file:

```{r}
my.genes <- rownames(TCGA_counts)
```
   
We calculate the exonic gene lenght for each gene contained in the the txdb file:

```{r}
exons.list.per.gene <- exonsBy(txdb,by="gene")
exonic.gene.sizes <- as.data.frame(sum(width(reduce(exons.list.per.gene))))
```

We then select the gene lenghts for the genes of insterest, that is my.genes:    
   
```{r}
exonic.gene.sizes <- exonic.gene.sizes[my.genes,]
```
  
And put the names back on the genes:

```{r}
names(exonic.gene.sizes) <- my.genes
```

Now we have the lenght for each gene in our table, this way we can compute the TPM data using the table of counts and the gene lenght:
    
```{r}
# michael's version
# [https://support.bioconductor.org/p/91218/](https://support.bioconductor.org/p/91218/)
tpm3 <- function(counts,len) {
 
  x <- counts/len
  x <- na.omit(x)
  
  return(t(t(x)*1e6/colSums(x)))
 
}
```
   
```{r}
TCGA_TPM <- as.data.frame(tpm3(as.matrix(TCGA_counts), exonic.gene.sizes))
```

```{r}
write.table(TCGA_TPM , "/Users/cristinacondel/Desktop/MAPK-MEK_Analysis/Gene_signatures/TCGA_TPM_ENSMBL.txt", sep = "\t", row.names = T, quote = F)
```